cpp/test/sg/tsne_test.cu

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef IF_DEBUG
+#define IF_DEBUG 1
+#endif
+
+#include <gtest/gtest.h>
+#include <score/scores.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <iostream>
+#include <vector>
+#include "tsne/digits.h"
+#include "tsne/tsne.cu"
+
+#include "cuda_utils.h"
+
+using namespace MLCommon;
+using namespace MLCommon::Score;
+using namespace MLCommon::Distance;
+
+using namespace ML;
+
+class TSNETest : public ::testing::Test {
```

Comment:
We should run both the barnes-hut and exact versions here, just to make sure we're testing both code paths.